### PR TITLE
op-deployer: Custom gas price estimator

### PR DESCRIPTION
**Pull request by mslipper**
Sepolia's fees are super high and extremely volatile right now. As a result, it became difficult for users to deploy new chains using op-deployer. The OPCM's deploy transaction buys most of the gas in the block, and the default gas price estimation logic in the transaction manager wasn't aggressive enough for the transactions to land on chain in a timely manner.

This PR adds the ability to customize the transaction manager with a custom `GasPriceEstimator` function that returns the tip, base fee, and blob fee. I extracted the original logic in the transaction manager into a default estimator that will be used if one isn't specified. For op-deployer, I built a custom estimator that pads the head block's base fee by 20%, and multiplies the suggested tip by 10. After testing this, I was able to get transactions onto Sepolia reliably. The algorithm is pretty simple and overpays for transactions by a lot, but for op-deployer's use case it's more important that transactions land quickly than it is they be cheap. Deployments are a one-time thing.

### CODE REVIEW

File: op-chain-ops/deployer/integration_test/apply_test.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ participants:
   - el_type: geth
     el_extra_params:
       - "--gcmode=archive"
+      - "--rpc.txfeecap=0"
```
Comment (axelKingsley) [inline op-chain-ops/deployer/integration_test/apply_test.go:32]: I don't see any other modifications to do with `txfeecap` in this PR -- is this change needed?

File: op-service/txmgr/cli.go [hunk 2]
```diff
@@ -418,6 +418,10 @@ type Config struct {
 	// Signer is used to sign transactions when the gas price is increased.
 	Signer opcrypto.SignerFn
 	From   common.Address
+
+	// GasPriceEstimator is used to estimate the gas price for a transaction.
+	// If nil, DefaultGasPriceEstimator is used.
+	GasPriceEstimator GasPriceEstimator
```
Comment (axelKingsley) [inline op-service/txmgr/cli.go]: ❤️ the abstraction!
Maybe include `Fn` on the type name, like the above "SignerFn"

File: op-service/txmgr/txmgr.go [hunk 3]
```diff
@@ -140,6 +140,7 @@ type SimpleTxManager struct {
 	backend ETHBackend
 	l       log.Logger
 	metr    metrics.TxMetricer
+	gpe     GasPriceEstimator
```
Comment (axelKingsley) [inline op-service/txmgr/txmgr.go]: can we call it `GasEstimationFunc` or something more verbose? `cfg` is a common shortening, but `gpe` is not familiar to me.

**Review by axelKingsley — COMMENTED** [review]
I like the abstraction, just some comments on naming.

And we could benefit from a test that confirms that the txmgr supports an overloaded estimation function. not sure if there's anything easy to extend for that.

File: op-chain-ops/deployer/integration_test/apply_test.go (hunk 1 above)
Comment (mslipper) [reply op-chain-ops/deployer/integration_test/apply_test.go:32]: Yes, it's necessary. The default fee cap in Geth is 1ETH, and the test goes over 1ETH when deploying contracts via the OPCM. This causes the transaction to be rejected and the test to fail.

**Review by axelKingsley — APPROVED** [review]

File: op-service/txmgr/txmgr.go [hunk 4]
```diff
@@ -876,27 +879,18 @@ func (m *SimpleTxManager) increaseGasPrice(ctx context.Context, tx *types.Transa
...
 		m.metr.RPCError()
-		return nil, nil, nil, fmt.Errorf("failed to fetch the suggested base fee: %w", err)
-	} else if head.BaseFee == nil {
-		return nil, nil, nil, errors.New("txmgr does not support pre-london blocks that do not have a base fee")
+		return nil, nil, nil, fmt.Errorf("failed to get gas price estimates: %w", err)
 	}
 
-	baseFee := head.BaseFee
-	m.metr.RecordBaseFee(baseFee)
-	m.metr.RecordTipCap(tip)
```
Comment (geoknee) [inline op-service/txmgr/txmgr.go:898]: Did you mean to remove these metric updates? `RecordTipCap` is now unreferenced / dead code.

Comment (sebastianst) [reply op-service/txmgr/txmgr.go:898]: Pretty sure this needs to be added back, both metrics.

File: op-service/txmgr/txmgr.go [hunk 5]
```diff
@@ -910,11 +904,6 @@ func (m *SimpleTxManager) SuggestGasPriceCaps(ctx context.Context) (*big.Int, *b
 		baseFee = new(big.Int).Set(minBaseFee)
 	}
 
-	var blobFee *big.Int
-	if head.ExcessBlobGas != nil {
-		blobFee = eip4844.CalcBlobFee(*head.ExcessBlobGas)
-		m.metr.RecordBlobBaseFee(blobFee)
```
Comment (sebastianst) [inline op-service/txmgr/txmgr.go:916]: this metric also needs to be added back.

**Review by sebastianst — COMMENTED** [review]
deleted metrics should be readded

File: op-service/txmgr/txmgr.go (hunk 4 above)
Comment (geoknee) [reply op-service/txmgr/txmgr.go:898]: https://github.com/ethereum-optimism/optimism/pull/12369
